deep `file:` relative dep not found

* package.json deps:
   + file:demo-snippets
* demo-snippets:
  + package.json deps:
     - file:../packages/collectionview
* packages/collectionview
   
<img width="1018" alt="Screenshot 2022-04-23 at 17 00 56" src="https://user-images.githubusercontent.com/655344/164911714-90c4fdcc-92da-419e-97d0-967b6f40bd28.png">

### pnpm version:

7.rc7

### Code to reproduce the issue:



### Expected behavior:

dependencies should be found

### Actual behavior:

```
 ERR_PNPM_LINKED_PKG_DIR_NOT_FOUND  Could not install from "/Volumes/dev/nativescript/packages/collectionview" as it does not exist.
 ```

### Additional information:

macos

--- a/.changeset/calm-lamps-occur.md
+++ b/.changeset/calm-lamps-occur.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/resolve-dependencies": patch
+"pnpm": patch
+---
+
+It should be possible to use a chain of local file dependencies [#4611](https://github.com/pnpm/pnpm/issues/4611).

--- a/packages/core/test/install/local.ts
+++ b/packages/core/test/install/local.ts
@@ -1,6 +1,7 @@
 import { promises as fs } from 'fs'
 import path from 'path'
 import { LOCKFILE_VERSION } from '@pnpm/constants'
+import { Lockfile } from '@pnpm/lockfile-file'
 import { prepareEmpty, preparePackages } from '@pnpm/prepare'
 import { addDistTag } from '@pnpm/registry-mock'
 import fixtures from '@pnpm/test-fixtures'
@@ -11,6 +12,7 @@ import {
 } from '@pnpm/core'
 import rimraf from '@zkochan/rimraf'
 import normalizePath from 'normalize-path'
+import readYamlFile from 'read-yaml-file'
 import symlinkDir from 'symlink-dir'
 import { testDefaults } from '../utils'
 
@@ -273,4 +275,7 @@ test('deep local', async () => {
   ])
   process.chdir('../project-1')
   await install(manifest1, await testDefaults())
+
+  const lockfile = await readYamlFile<Lockfile>('pnpm-lock.yaml')
+  expect(Object.keys(lockfile.packages ?? {})).toStrictEqual(['file:../project-2', 'file:../project-2/project-3'])
 })

--- a/packages/core/test/install/local.ts
+++ b/packages/core/test/install/local.ts
@@ -1,7 +1,7 @@
 import { promises as fs } from 'fs'
 import path from 'path'
 import { LOCKFILE_VERSION } from '@pnpm/constants'
-import { prepareEmpty } from '@pnpm/prepare'
+import { prepareEmpty, preparePackages } from '@pnpm/prepare'
 import { addDistTag } from '@pnpm/registry-mock'
 import fixtures from '@pnpm/test-fixtures'
 import {
@@ -238,4 +238,39 @@ test('frozen-lockfile: installation fails if the integrity of a tarball dependen
   await expect(
     install(manifest, await testDefaults({ frozenLockfile: true }))
   ).rejects.toThrow(/Got unexpected checksum/)
+})
+
+test('deep local', async () => {
+  const manifest1 = {
+    name: 'project-1',
+    version: '1.0.0',
+    dependencies: {
+      'project-2': 'file:../project-2',
+    },
+  }
+  preparePackages([
+    {
+      location: 'project-1',
+      package: manifest1,
+    },
+    {
+      location: 'project-2',
+      package: {
+        name: 'project-2',
+        version: '1.0.0',
+        dependencies: {
+          'project-3': 'file:./project-3',
+        },
+      },
+    },
+    {
+      location: 'project-2/project-3',
+      package: {
+        name: 'project-3',
+        version: '1.0.0',
+      },
+    },
+  ])
+  process.chdir('../project-1')
+  await install(manifest1, await testDefaults())
 })

--- a/packages/resolve-dependencies/src/resolveDependencies.ts
+++ b/packages/resolve-dependencies/src/resolveDependencies.ts
@@ -836,7 +836,7 @@ async function resolveDependency (
     normalizedPref: options.currentDepth === 0 ? pkgResponse.body.normalizedPref : undefined,
     pkgId: pkgResponse.body.id,
     prefix: pkgResponse.body.resolution.type === 'directory'
-      ? path.resolve(ctx.prefix, pkgResponse.body.resolution['directory'])
+      ? path.resolve(ctx.lockfileDir, pkgResponse.body.resolution['directory'])
       : ctx.prefix,
 
     // Next fields are actually only needed when isNew = true

--- a/packages/resolve-dependencies/src/resolveDependencies.ts
+++ b/packages/resolve-dependencies/src/resolveDependencies.ts
@@ -159,6 +159,7 @@ export type PkgAddress = {
   pkg: PackageManifest
   version?: string
   updated: boolean
+  prefix: string
 } & ({
   isLinkedDependency: true
   version: string
@@ -320,7 +321,10 @@ async function resolveDependenciesOfDependency (
 
   postponedResolutionsQueue.push(async (preferredVersions) =>
     resolveChildren(
-      ctx,
+      {
+        ...ctx,
+        prefix: resolveDependencyResult.prefix,
+      },
       resolveDependencyResult,
       extendedWantedDep.infoFromLockfile?.dependencyLockfile,
       options.workspacePackages,
@@ -831,6 +835,9 @@ async function resolveDependency (
     nodeId,
     normalizedPref: options.currentDepth === 0 ? pkgResponse.body.normalizedPref : undefined,
     pkgId: pkgResponse.body.id,
+    prefix: pkgResponse.body.resolution.type === 'directory'
+      ? path.resolve(ctx.prefix, pkgResponse.body.resolution['directory'])
+      : ctx.prefix,
 
     // Next fields are actually only needed when isNew = true
     installable,

--- a/packages/resolve-dependencies/src/resolveDependencies.ts
+++ b/packages/resolve-dependencies/src/resolveDependencies.ts
@@ -159,7 +159,7 @@ export type PkgAddress = {
   pkg: PackageManifest
   version?: string
   updated: boolean
-  prefix: string
+  rootDir: string
 } & ({
   isLinkedDependency: true
   version: string
@@ -200,7 +200,7 @@ export interface ResolvedPackage {
   }
 }
 
-type ParentPkg = Pick<PkgAddress, 'nodeId' | 'installable' | 'depPath'>
+type ParentPkg = Pick<PkgAddress, 'nodeId' | 'installable' | 'depPath' | 'rootDir'>
 
 interface ResolvedDependenciesOptions {
   currentDepth: number
@@ -321,10 +321,7 @@ async function resolveDependenciesOfDependency (
 
   postponedResolutionsQueue.push(async (preferredVersions) =>
     resolveChildren(
-      {
-        ...ctx,
-        prefix: resolveDependencyResult.prefix,
-      },
+      ctx,
       resolveDependencyResult,
       extendedWantedDep.infoFromLockfile?.dependencyLockfile,
       options.workspacePackages,
@@ -629,7 +626,7 @@ async function resolveDependency (
         !wantedDependency.pref.startsWith('file:')
       )
         ? ctx.lockfileDir
-        : ctx.prefix,
+        : options.parentPkg.rootDir,
       registry: wantedDependency.alias && pickRegistryForPackage(ctx.registries, wantedDependency.alias, wantedDependency.pref) || ctx.registries.default,
       // Unfortunately, even when run with --lockfile-only, we need the *real* package.json
       // so fetching of the tarball cannot be ever avoided. Related issue: https://github.com/pnpm/pnpm/issues/1176
@@ -827,6 +824,9 @@ async function resolveDependency (
     }
   }
 
+  const rootDir = pkgResponse.body.resolution.type === 'directory'
+    ? path.resolve(ctx.lockfileDir, pkgResponse.body.resolution['directory'])
+    : ctx.prefix
   return {
     alias: wantedDependency.alias || pkg.name,
     depIsLinked,
@@ -835,9 +835,7 @@ async function resolveDependency (
     nodeId,
     normalizedPref: options.currentDepth === 0 ? pkgResponse.body.normalizedPref : undefined,
     pkgId: pkgResponse.body.id,
-    prefix: pkgResponse.body.resolution.type === 'directory'
-      ? path.resolve(ctx.lockfileDir, pkgResponse.body.resolution['directory'])
-      : ctx.prefix,
+    rootDir,
 
     // Next fields are actually only needed when isNew = true
     installable,

--- a/packages/resolve-dependencies/src/resolveDependencyTree.ts
+++ b/packages/resolve-dependencies/src/resolveDependencyTree.ts
@@ -133,6 +133,7 @@ export default async function<T> (
         nodeId: `>${importer.id}>`,
         optional: false,
         depPath: importer.id,
+        rootDir: importer.rootDir,
       },
       proceed,
       resolvedDependencies: {

